From 8bda3dad98f3e1d62bf4499ecb95e2692830adaa Mon Sep 17 00:00:00 2001
From: Jason Gerecke <killertofu@gmail.com>
Date: Thu, 10 Aug 2017 15:35:43 -0700
Subject: [PATCH 4/4] Correct device flags for multiple devices

The device flags set in wcmDeviceTypeKeys must be set properly in order
for the driver to correctly identify and use specific tablet features.
Several devices are missing flags that e.g. indicate they support
pen rotation or have touch rings.

Fixes: 0318b1c ("Support recent display devices on older kernels")
Fixes: 3578b91 ("Add support for Cintiq 22HDT and 13HD, DTK2241 and DTH2242.")
Fixes: df7f165 ("Add tablet_type to WacomModel")
Signed-off-by: Jason Gerecke <jason.gerecke@wacom.com>
Reviewed-by: Ping Cheng <ping.cheng@wacom.com>
(cherry picked from commit b30e2d27fdb30ff4b425fc62432cab636fe163b9)
---
 src/wcmValidateDevice.c | 20 +++++++++++++-------
 1 file changed, 13 insertions(+), 7 deletions(-)

diff --git a/src/wcmValidateDevice.c b/src/wcmValidateDevice.c
index 8eac0dd..21ccd5f 100644
--- a/src/wcmValidateDevice.c
+++ b/src/wcmValidateDevice.c
@@ -199,7 +199,12 @@ int wcmDeviceTypeKeys(InputInfoPtr pInfo)
 	{
 		case 0xF8:  /* Cintiq 24HDT */
 		case 0xF4:  /* Cintiq 24HD */
-			TabletSetFeature(priv->common, WCM_DUALRING | WCM_LCD);
+			TabletSetFeature(priv->common, WCM_DUALRING);
+			/* fall through */
+
+		case 0x34D: /* MobileStudio Pro 13 */
+		case 0x34E: /* MobileStudio Pro 13 */
+			TabletSetFeature(priv->common, WCM_LCD);
 			/* fall through */
 
 		case 0x357: /* Intuos Pro 2 M */
@@ -300,23 +305,24 @@ int wcmDeviceTypeKeys(InputInfoPtr pInfo)
 		case 0xED: /* TPC with 1FGT */
 		case 0x90: /* TPC */
 		case 0x97: /* TPC */
+		case 0x9F: /* TPC */
 		case 0xEF: /* TPC */
 			TabletSetFeature(priv->common, WCM_TPC);
 			break;
 
-		case 0x9F:
+		case 0x304:/* Cintiq 13HD */
+		case 0x34F:/* Cintiq Pro 13 FHD */
+		case 0x350:/* Cintiq Pro 16 UHD */
+			TabletSetFeature(priv->common, WCM_ROTATION);
+			/* fall-through */
+
 		case 0xF6: /* Cintiq 24HDT Touch */
 		case 0x57: /* DTK2241 */
 		case 0x59: /* DTH2242 Pen */
 		case 0x5D: /* DTH2242 Touch */
 		case 0x5E: /* Cintiq 22HDT Touch */
-		case 0x304:/* Cintiq 13HD */
 		case 0x34A:/* MobileStudio Pro 13 Touch */
 		case 0x34B:/* MobileStudio Pro 16 Touch */
-		case 0x34D:/* MobileStudio Pro 13 */
-		case 0x34E:/* MobileStudio Pro 13 */
-		case 0x34F:/* Cintiq Pro 13 FHD */
-		case 0x350:/* Cintiq Pro 16 UHD */
 		case 0x353:/* Cintiq Pro 13 FHD Touch */
 		case 0x354:/* Cintiq Pro 13 UHD Touch */
 		case 0x35A:/* DTH-1152*/
-- 
2.13.6

